conn.go

```diff
@@ -20,34 +23,110 @@ var log = logging.Logger("conn")
 
 // ReleaseBuffer puts the given byte array back into the buffer pool,
 // first verifying that it is the correct size
+// TODO: do we need this? I don't see it's used anywhere
 func ReleaseBuffer(b []byte) {
 	log.Debugf("Releasing buffer! (cap,size = %d, %d)", cap(b), len(b))
 	mpool.ByteSlicePool.Put(uint32(cap(b)), b)
 }
 
-// singleConn represents a single connection to another Peer (IPFS Node).
+// singleConn represents a single stream-multipexed connection to another Peer (IPFS Node).
 type singleConn struct {
-	local  peer.ID
-	remote peer.ID
-	maconn tpt.Conn
-	event  io.Closer
-}
+	streamConn smux.Conn
+	tptConn    tpt.Conn
+
+	secSession secio.Session
+
+	event io.Closer
+}
+
+var _ iconn.Conn = &singleConn{}
+
+// newSingleConn constructs a new connection
+func newSingleConn(ctx context.Context, local, remote peer.ID, privKey ci.PrivKey, tptConn tpt.Conn, pstpt smux.Transport, isServer bool) (iconn.Conn, error) {
+	ml := lgbl.Dial("conn", local, remote, tptConn.LocalMultiaddr(), tptConn.RemoteMultiaddr())
+
+	var streamConn smux.Conn
+	var secSession secio.Session
+	switch conn := tptConn.(type) {
+	case tpt.SingleStreamConn:
+		c := conn
+		// 1. secure the connection
+		if iconn.EncryptConnections {
+			secSession, err := setupSecureSession(ctx, local, privKey, conn)
+			if err != nil {
+				// TODO: return a more specific error code here
+				return nil, err
+			}
+			c = &secureConn{
+				SingleStreamConn: conn,
+				secure:           secSession,
+			}
+		}
 
-// newConn constructs a new connection
-func newSingleConn(ctx context.Context, local, remote peer.ID, maconn tpt.Conn) (iconn.Conn, error) {
-	ml := lgbl.Dial("conn", local, remote, maconn.LocalMultiaddr(), maconn.RemoteMultiaddr())
+		// 2. start stream multipling
+		var err error
+		streamConn, err = pstpt.NewConn(c, isServer)
+		if err != nil {
+			return nil, err
+		}
+	case tpt.MultiStreamConn:
+		// 1. secure the connection on the first best stream
```

Comment:
I think we want to add a big TODO and WARNING here, we're only doing this this way for now, for quic, until we have the ability to simply rely on QUICs crypto to validate the peers. Its also insecure (can be man in the middled) so we should note that QUIC is experimental for now.